File: 3606-coupon-code-validator/3606-coupon-code-validator.java
class Solution {
    public List<String> validateCoupons(String[] code, String[] businessLine, boolean[] isActive) {

        Map<String, List<String>> map = new HashMap<>();
        map.put("electronics", new ArrayList<>());
        map.put("grocery", new ArrayList<>());
        map.put("pharmacy", new ArrayList<>());
        map.put("restaurant", new ArrayList<>());

        for (int i = 0; i < code.length; i++) {
            if (
                isActive[i] &&
                code[i].matches("[a-zA-Z0-9_]+") &&
                map.containsKey(businessLine[i])
            ) {
                map.get(businessLine[i]).add(code[i]);
            }
        }

        List<String> ans = new ArrayList<>();

        String[] order = {"electronics", "grocery", "pharmacy", "restaurant"};
        for (String b : order) {
            Collections.sort(map.get(b));
            ans.addAll(map.get(b));
        }

        return ans;
    }
}

File: 3606-coupon-code-validator/3606-coupon-code-validator.cpp
class Solution {
public:
    vector<string> validateCoupons(vector<string>& code, vector<string>& businessLine, vector<bool>& isActive) {
        int n = code.size();
        vector<pair<string,string>>temp;
        vector<string>ans;
        for(int i=0;i<n;i++){
            if(isActive[i] && code[i]!="" && (businessLine[i]== "electronics"||businessLine[i]== "grocery" || businessLine[i]==
            "pharmacy" || businessLine[i]=="restaurant")){
                bool flag = true;
                for(char ch: code[i]){
                    if(!(isalnum(ch)|| ch=='_')){
                        flag  = false;
                    }
                }
                if(flag) temp.push_back({businessLine[i],code[i]});
            }
        }
        sort(temp.begin(), temp.end());

        for(auto& tem : temp){
            ans.push_back(tem.second);
        }
        return ans;
    }
};